src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -639,38 +619,45 @@ std::pair<bool, Uptane::Target> SotaUptaneClient::downloadImage(const Uptane::Ta
   }
 
-  KeyManager keys(storage, config.keymanagerConfig());
-  keys.loadKeys();
-  auto prog_cb = [this](const Uptane::Target &t, const std::string &description, unsigned int progress) {
-    report_progress_cb(events_channel.get(), t, description, progress);
-  };
-
   bool success = false;
-  const Uptane::EcuSerial &primary_ecu_serial = primaryEcuSerial();
-
-  if (target.IsForEcu(primary_ecu_serial) || !target.IsOstree()) {
-    // TODO: download should be the logical ECU and packman specific
-    const int max_tries = 3;
-    int tries = 0;
-    std::chrono::milliseconds wait(500);
-
-    for (; tries < max_tries; tries++) {
-      success = package_manager_->fetchTarget(target, *uptane_fetcher, keys, prog_cb, token);
-      // Skip trying to fetch the 'target' if control flow token transaction
-      // was set to the 'abort' or 'pause' state, see the CommandQueue and FlowControlToken.
-      if (success || (token != nullptr && !token->canContinue(false))) {
-        break;
-      } else if (tries < max_tries - 1) {
-        std::this_thread::sleep_for(wait);
-        wait *= 2;
+  try {
+    KeyManager keys(storage, config.keymanagerConfig());
+    keys.loadKeys();
+    auto prog_cb = [this](const Uptane::Target &t, const std::string &description, unsigned int progress) {
+      report_progress_cb(events_channel.get(), t, description, progress);
+    };
+
+    const Uptane::EcuSerial &primary_ecu_serial = primaryEcuSerial();
+
+    if (target.IsForEcu(primary_ecu_serial) || !target.IsOstree()) {
+      // TODO: download should be the logical ECU and packman specific
+      const int max_tries = 3;
+      int tries = 0;
+      std::chrono::milliseconds wait(500);
+
+      for (; tries < max_tries; tries++) {
+        success = package_manager_->fetchTarget(target, *uptane_fetcher, keys, prog_cb, token);
+        // Skip trying to fetch the 'target' if control flow token transaction
+        // was set to the 'abort' or 'pause' state, see the CommandQueue and FlowControlToken.
+        if (success || (token != nullptr && !token->canContinue(false))) {
+          break;
+        } else if (tries < max_tries - 1) {
+          std::this_thread::sleep_for(wait);
+          wait *= 2;
+        }
       }
+      if (!success) {
+        LOG_ERROR << "Download unsuccessful after " << tries << " attempts.";
+        // TODO: show real exception
+        throw Uptane::TargetHashMismatch(target.filename());
+      }
+    } else {
+      // we emulate successfull download in case of the Secondary OSTree update
+      success = true;
     }
-    if (!success) {
-      LOG_ERROR << "Download unsuccessful after " << tries << " attempts.";
-    }
-  } else {
-    // we emulate successfull download in case of the Secondary OSTree update
-    success = true;
+  } catch (const std::exception &e) {
+    LOG_ERROR << "Error downloading image: " << e.what();
+    last_exception = std::current_exception();
```

Comment:
What prevents us from moving this try..catch block one level higher, to the `downloadImages` method and get rid of `success` return value?

Reply:
Because of the EcuDownloadCompleteReport and DownloadTarget event. I tried to move them out but then noticed that aktualizr-lite is calling `downloadImage()` directly, so it would miss these calls right now.

We'd have to make some changes there to fix this. I can add a comment.

Reply:
Right, I missed them somehow.